src/nfa/nfa-intersection.cc

```diff
@@ -119,27 +120,51 @@ class Intersection
     /**
      * Compute intersection preserving epsilon transitions.
      */
-    void compute_preserving_epsilon_transitions()
-    {
+    void compute_preserving_epsilon_transitions() {
         initialize_pairs_to_process();
 
-        if (lhs.trans_empty() || rhs.trans_empty()) { return; }
-
-        while (!pairs_to_process.empty())
-        {
+        while (!pairs_to_process.empty()) {
             pair_to_process = *pairs_to_process.begin();
             pairs_to_process.erase(pair_to_process);
+            compute_transitions_for_eps_pres_prod_for_state_pair();
+        }
+    }
 
-            // find all transitions that have same symbol for first and the second state in the pair_to_process
-            for (const auto& this_state_transitions: lhs.transitionrelation[pair_to_process.first] )
-            {
-                for (const auto& other_state_transitions: rhs.transitionrelation[pair_to_process.second])
-                {
-                    compute_for_state_pair_preserving_eps_trans(this_state_transitions,
-                                                                other_state_transitions);
+    /**
+     * Compute transitions for epsilon preserving product for a current state pair.
+     */
+    void compute_transitions_for_eps_pres_prod_for_state_pair() {
+        for (const auto& lhs_state_transitions: lhs.transitionrelation[pair_to_process.first]) {
+            // Check for lhs epsilon transitions.
+            if (lhs_state_transitions.symbol == epsilon) {
+                compute_for_lhs_state_epsilon_transitions(lhs_state_transitions);
+            }
+
+            for (const auto& rhs_state_transitions: rhs.transitionrelation[pair_to_process.second]) {
+                // Check for rhs epsilon transitions.
+                if (rhs_state_transitions.symbol == epsilon) {
+                    compute_for_rhs_state_epsilon_transitions(rhs_state_transitions);
+                }
+
+                // Find all transitions that have the same symbol for first and the second state in the pair_to_process.
+                if (lhs_state_transitions.symbol == rhs_state_transitions.symbol) {
+                    compute_for_same_symbols(lhs_state_transitions, rhs_state_transitions);
                 }
             }
         }
+
+        compute_for_independent_rhs_epsilon_transitions();
+    }
+
+    /**
+     * Check for epsilon transitions in case only rhs has any transitions.
+     */
+    void compute_for_independent_rhs_epsilon_transitions() {
```

Comment:
Wouldn't be `find_epsilon_trans_in_rhs`  a better namae?

Reply:
The function is **computing intersection** for **epsilon transitions** in NFA rhs if lhs has no transitions. Otherwise, rhs epsilon transitions would be skipped and not **added** to the intersection.

Hence, to reflect the actions of the function, the name should express the following: adding the remaining rhs epsilon transitions to the intersection. So, maybe `add_rhs_epsilon_transitions` could be better. What do you think?

Reply:
`add_rhs_epsilon_transitions` is a good name.

Reply:
Fixed.